modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/queue/ViewableQueue.scala

```diff
@@ -0,0 +1,51 @@
+package org.tessellation.node.shared.domain.queue
+
+import cats.effect.std.{Mutex, Queue}
+import cats.effect.{Concurrent, Ref}
+import cats.implicits._
+
+trait ViewableQueue[F[_], A] {
+  def offer(item: A): F[Unit]
+
+  def take: F[A]
+
+  def tryOfferN(items: List[A]): F[List[A]]
+
+  def tryTakeN(n: Option[Int]): F[List[A]]
+
+  def view: F[List[A]]
+}
+
+object ViewableQueue {
+
+  def unbounded[F[_]: Concurrent, A]: F[ViewableQueue[F, A]] =
+    Queue.unbounded[F, A].flatMap(q => make(q))
+
+  def make[F[_]: Concurrent, A](queue: Queue[F, A]): F[ViewableQueue[F, A]] =
+    (
+      Ref[F].of(List.empty[A]),
+      Mutex[F]
+    ).tupled.map {
+      case (state, mutex) =>
+        new ViewableQueue[F, A] {
+          def offer(item: A): F[Unit] =
+            mutex.lock.use(_ => queue.offer(item) *> state.update(_ :+ item))
+
+          def take: F[A] =
+            mutex.lock.use(_ => queue.take.flatTap(item => state.update(_.filterNot(_ == item))))
+
+          def tryOfferN(items: List[A]): F[List[A]] =
+            mutex.lock.use { _ =>
+              queue.tryOfferN(items).flatMap { failedItems =>
+                val successfulItems = items.diff(failedItems)
+                state.update(_ ++ successfulItems).as(failedItems)
+              }
+            }
+
+          def tryTakeN(n: Option[Int]): F[List[A]] =
+            mutex.lock.use(_ => queue.tryTakeN(n).flatTap(taken => state.update(cur => cur.diff(taken))))
+
+          def view: F[List[A]] = state.get
```

Comment:
Views may be out of date if queried while the mutex is locked but my intention is to minimize the impact to write operations.